src/types/external-source.ts

```diff
@@ -0,0 +1,81 @@
+import type { ExternalEventDB, ExternalEventInsertInput, ExternalEventJson } from '../types/external-event';
+import type { UserId } from './app';
+
+// Represents all fields used as a composite primary key for merlin.external_source
+export type ExternalSourcePkey = {
+  derivation_group_name: string;
+  key: string;
+};
+
+// This is the type that conforms with the database schema. We don't really use it, as it is pretty heavyweight - instead we derive lighter types from it.
+export type ExternalSourceDB = {
+  created_at: string;
+  derivation_group_name: string;
+  end_time: string;
+  external_events: ExternalEventDB[];
+  key: string;
+  metadata: Record<string, any>;
+  owner: UserId;
+  source_type_name: string;
+  start_time: string;
+  valid_at: string;
+};
+
+// This is the JSON type that the user can upload.
+export type ExternalSourceJson = {
+  events: ExternalEventJson[];
+  source: {
+    key: string;
+    metadata: object;
+    period: {
+      end_time: string;
+      start_time: string;
+    };
+    source_type: string;
+    valid_at: string;
+  };
+};
+
+// For use in retrieval of source information sans bulky items like metadata and event lists (see stores)
+export type ExternalSourceSlim = Omit<ExternalSourceDB, 'external_events' | 'metadata'>;
+
+// Similar to ExternalSourceDB, but uses ExternalSourcePkey to represent the primary key (key, derivation_group_name)
+export type ExternalSource = Omit<ExternalSourceDB, 'key' | 'derivation_group_name'> & { pkey: ExternalSourcePkey };
+
+export type PlanDerivationGroup = {
+  acknowledged: boolean;
+  derivation_group_name: string;
+  last_acknowledged_at: string;
+  plan_id: number | undefined; // because in plan.ts plan is defined on Plan | null...
```

Comment:
I _think_ it might be in reference to the `plan` readable store itself being typed to `Plan | null` - but I don't think its necessary as well as typing the plan_id to possibly undefined, removed the comment and set `plan_id: number`.